### PR TITLE
Feature/rb 152 upgrade bullet

### PR DESCRIPTION
**Pull request by romainpanno**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Norm of the code has been respected

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of change does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [x] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **What is the current behavior?** (link an issue based on the kind of change this pr introduce)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

## Summary by CodeRabbit

New Features:
- Introduced a new missile system, replacing the previous bullet system for enhanced gameplay.
- Added missile health and missile death features, providing more interactive and dynamic combat scenarios.
- Implemented a missile death notification system, improving real-time game updates.

Refactor:
- Updated various function names and parameters for better code clarity and consistency.
- Optimized network message handling for missile-related events, improving network efficiency.

Bug Fixes:
- Fixed potential issues with missile creation and tracking, ensuring smoother gameplay.

Documentation:
- Updated RFC documentation to reflect the new missile system and related changes.

### CODE REVIEW

**Review by kw4ntiq — APPROVED** [review]

File: docs/network/rfc/RFC.md [hunk 1]
```diff
@@ -423,6 +427,22 @@ Table of Contents
     This field must be of size 4 bytes.
     This field is signed (so starting from -((2^32)/2) to +(((2^32)/2)-1))
 
+    *** Missile ID
+    
+    This field correspond to the ID of the missile.
+    This is useless because this is the server that will set the ID of the
+    missile.
+    
+    This field must be of size 4 byte.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:436]:
```suggestion
    This field must be of size 4 byte.
    This field must be equal to `0`
```

File: docs/network/rfc/RFC.md [hunk 2]
```diff
@@ -423,6 +427,22 @@ Table of Contents
...
+    
+    *** Missile Health
+    
+    This field correspond to the health of the missile.
+    This is useless because this is the server that will set the health of the
+    missile.
+    
+    This field must be of size 4 byte.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:445]:
```suggestion
    This field must be of size 4 byte.
    This field must be equal to `0`
```

File: docs/network/rfc/RFC.md [hunk 3]
```diff
@@ -515,11 +535,34 @@ Table of Contents
...
+   
+    This field help to know the packet is realy missile death action
+    
+    This field must be of size 1 byte.
+    This field is unsigned (so starting from 0 to 2^8)
+    This field must be equal to the ascii `\x16`
+    
+    *** Missile ID
+    
+    This field correspond to the ID of the missile that died
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:558]:
```suggestion
    This field correspond to the ID of the missile that died

```

File: docs/network/rfc/RFC.md [hunk 4]
```diff
@@ -852,6 +897,16 @@ Table of Contents
     This field must be of size 4 bytes.
     This field is signed (so starting from -((2^32)/2) to +(((2^32)/2)-1))
 
+    *** Missile ID
+    
+    This field correspond to the ID of the missile.
+    This field must be of size 4 bytes.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:907]: signed ? unsigned ?
this field is uniq for each missile ?

File: docs/network/rfc/RFC.md [hunk 5]
```diff
@@ -852,6 +897,16 @@ Table of Contents
     This field must be of size 4 bytes.
     This field is signed (so starting from -((2^32)/2) to +(((2^32)/2)-1))
 
+    *** Missile ID
+    
+    This field correspond to the ID of the missile.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:905]:
```suggestion
    This field correspond to the ID of the missile.

```

File: docs/network/rfc/RFC.md [hunk 6]
```diff
@@ -852,6 +897,16 @@ Table of Contents
...
+    This field must be of size 4 bytes.
+    
+    *** Missile Health
+    
+    This field correspond to the health of the missile.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:912]:
```suggestion
    This field correspond to the health of the missile.

```

File: docs/network/rfc/RFC.md [hunk 7]
```diff
@@ -852,6 +897,16 @@ Table of Contents
...
+    This field must be of size 4 bytes.
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:914]: signed ? unsigned ?

File: docs/network/rfc/RFC.md [hunk 8]
```diff
@@ -972,11 +1027,34 @@ Table of Contents
...
+   
+    This field help to know the packet is realy missile death action
+    
+    This field must be of size 1 byte.
+    This field is unsigned (so starting from 0 to 2^8)
+    This field must be equal to the ascii `\x16`
+    
+    *** Missile ID
+    
+    This field correspond to the ID of the missile that died
```
Comment (Saverio976) [inline docs/network/rfc/RFC.md:1055]:
```suggestion
    This field correspond to the ID of the missile that died

```

File: src/Client/Systems/Network/ClientNetwork.cpp [hunk 9]
```diff
@@ -49,6 +49,28 @@ namespace Systems {
...
+        auto &arrHealth              = Registry::getInstance().getComponents<struct health_s>();
+        std::vector<std::size_t> ids = missiles.getExistingsId();
+
+        for (auto id : ids) {
+            if (missiles[id].constId == missileDeath.missileId) {
+                if (arrHealth.exist(id)) {
+                    arrHealth[id].hp = 0;
+                } else {
+                    Logger::fatal("\n\n\n!!!! Missile has no health component, but is alive !!!!\n\n\n");
+                    // TODO : remove missile
```
Comment (Saverio976) [inline src/Client/Systems/Network/ClientNetwork.cpp:66]: TODO ?

File: src/ECS/ECSCustomTypes.hpp [hunk 10]
```diff
@@ -83,7 +83,30 @@ namespace Types {
     };
 
     struct Missiles {
+        public:
+            Missiles(enum missileTypes_e _type = missileTypes_e::CLASSIC) : type(_type), constId(0)
```
Comment (Saverio976) [inline src/ECS/ECSCustomTypes.hpp]: alors, enum missileType_e c'est du C, le C n'a pas de namespacen et le fait que ça compile dit juste que ça va foutre du caca

Comment (Saverio976) [reply src/ECS/ECSCustomTypes.hpp]: CLASSIC

Comment (Saverio976) [reply src/ECS/ECSCustomTypes.hpp]:
```suggestion
            Missiles(enum missileTypes_e _type = CLASSIC) : type(_type), constId(0)
```

File: src/Nitwork/NitworkServer.hpp [hunk 11]
```diff
@@ -106,9 +107,7 @@ namespace Nitwork {
              * @param msg The infos of the bullet that will be created
              * @param senderEndpoint The endpoint of the client that sent the msg of hie new bullet
```
Comment (Saverio976) [inline src/Nitwork/NitworkServer.hpp]:
```suggestion
```

**Review by Saverio976 — CHANGES_REQUESTED** [review]

File: src/Client/Systems/Network/ClientNetwork.cpp (hunk 9 above)
Comment (romainpanno) [reply src/Client/Systems/Network/ClientNetwork.cpp:66]: il faudra utiliser le fait d'ajouter à la liste des components à remove qui est sur dev

File: src/ECS/ECSCustomTypes.hpp [hunk 12]
```diff
@@ -83,7 +83,30 @@ namespace Types {
...
+                return _missileNb;
+            }
+
             missileTypes_e type;
+            unsigned int constId;
+
+        private:
+            static unsigned int _missileNb;
+            static std::mutex _mutex;
     };
```
[the remark this answers is not in the document]
Comment (romainpanno) [reply src/ECS/ECSCustomTypes.hpp:110]: @coderabbitai no because every times i manipulate _missileNb, i use the mutex in the methods

**Comment by Saverio976** [comment]
https://github.com/X-R-G-B/R-Bus/pull/143

**Pull request closed without merging**